main/src/library/MutList.flix

```diff
@@ -833,6 +871,8 @@ namespace MutList {
     ///
     /// Applies `f` to all the elements in `v`.
     ///
+    /// pub def foreachWithIndex(f: (Int32, a) -> Unit \ { ef - Write(r) }, v: MutList[a, r]): Unit \ { ef, Read(r) }
+    ///
```

Comment:
I would like to see the rest of the file annotated. In particular, I am interested in if the private functions change anything.

(Maybe not, if none of them are higher-order, but what if they were?)

Reply:
Looking through the file, most private functions appear in the shape of `loop`. I have not annotated all of these cases, so let me summarize the most interesting.

---
__General Case:__  
Consider the following from the code (`map` to be precise):
```scala
def loop(i) = { // Int32 -> Unit \ { ef, Read(r), Write(r) }
                if (i >= l)
                    ()
                else {
                    b[i] = f(a[i]); // { ef, Read(r), Write(r) }
                    loop(i + 1)
                }
            }
```
`loop` can be thought of as a higher order function under eta-expansion (binding `f` which is an argument to the `map` function) which would have the type `(a -> b \ { { ef - Write(r) }, Read(r) }) -> Int32 -> Unit \ { ef, Read(r), Write(r) }`.

---
__Continuation__  
Consider the `loop` from `foldRightWithCont`:
```scala
def loop(i) = { // { ef, Read(r) }
            if (i >= l)
                z
            else {
                f(a[i], _ -> loop(i + 1)) // { ef, Read(r) }
            }
        }
```
I do not think this case, in and of itself is that interesting, since we can require `f` not having a `Write(r)` effect. But imagine a `mapWithCont`(whatever that would be). If we base it on the previous example, `loop` would have a `Write(r)` effect - but how do we then prevent `f` from performing writes outside of invocation of `loop`?  
Doing the `{ ef - Write(r) }` would not cut it, since the function may invoke `loop`.

---
__Iterator:__  
Lastly, let us look at `iterator`
```scala
    pub def iterator(l: MutList[a, r]): Iterator[a, r] \ { Write(r) } =
        let r = Scoped.regionOf(l);
        let MutList(a, len) = l;
        let i = ref 0 @ r; // Write(r)
        let done = () -> deref i == deref len; // Read(r)
        let next = () -> { // { Read(r), Write(r) }
            let x = (deref a)[deref i]; // Read(r)
            i := deref i + 1;  // { Read(r), Write(r) }
            x 
        };
        Iterator(done, next)
```
I we discussed, I find `next` having a `{ Read(r), Write(r) }` effect to be somewhat misleading,
given the `Write(r)` does not actually affect the underlying `MutList`, but rather the `i` variable, which might as well be allocated in a different region.  
This consideration of course, does not depend on effect restriction, so should be expressible already.